### PR TITLE
feat: add interface validation method for dynamic link functions

### PR DESCRIPTION
**Pull request by loloicci**
# Description
Add `validate_dynamic_link_interface` to `deps.api`.
contracts/dynamic-caller-contract/src/contract.rs is an example of how to use this.

## Types of changes

- [ ] Bug fix (changes which fixes an issue)
- [x] New feature (changes which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ETC (build, ci, docs, perf, refactor, style, test)

## Checklist

- [ ] I followed the [contributing guidelines](https://github.com/line/cosmwasm/blob/main/CONTRIBUTING.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] The PR title and commits are followed [conventional commit form](https://www.conventionalcommits.org/en/v1.0.0).

### CODE REVIEW

**Review by zemyblue — COMMENTED** [review]
@loloicci, the method that currently implemented interface detection function in this dynamic link is not rustLang specialized, is it? In other words, if smart contract can be developed in other development languages, it can be linked, right?

**Comment by loloicci** [comment]
> the method that currently implemented interface detection function in this dynamic link is not rustLang specialized, is it? In other words, if smart contract can be developed in other development languages, it can be linked, right?

Exactry, yes.

**Comment by shiki-tak** [comment]
As for the design, LGTM
In addition to validating whether an arbitrary function is implemented, it would be better to have the ability to detect whether a particular function is implemented and what kind of token it is, as in ERC 165.

**Comment by loloicci** [comment]
>  it would be better to have the ability to detect whether a particular function is implemented and what kind of token it is, as in ERC 165.

It looks good, but this PR is only for interface validation. So I will leave it on another issue.

**Review by zemyblue — COMMENTED** [review]
How can I check the code coverage?

**Comment by loloicci** [comment]
> How can I check the code coverage?
now we do not have features to check code coverage...

File: contracts/dynamic-caller-contract/src/contract.rs [hunk 1]
```diff
@@ -144,6 +150,26 @@ pub fn try_do_panic(deps: DepsMut, _env: Env) -> Result<Response, ContractError>
     Ok(Response::default())
 }
 
+pub fn try_validate_interface(deps: Deps, _env: Env) -> Result<Response, ContractError> {
+    let address = from_slice(&deps.storage.get(b"dynamic_callee_contract").unwrap())?;
```
[the remark this answers is not in the document]
Comment (loloicci) [reply contracts/dynamic-caller-contract/src/contract.rs:154]: in the line https://github.com/line/cosmwasm/pull/245/files/045ac42de920cefa0688b4017010a675595678da#diff-034266d0dbc7e843b378e05bd3c7c3c98c9435e81a518bbe020615543c782e7dR155, `address` is used as `Addr`. So, the compiler inference the type of `address` as `Addr`.

File: packages/vm/src/dynamic_link.rs [hunk 2]
```diff
@@ -234,6 +236,56 @@ where
...
+    interface: u32,
+) -> Result<u32, RuntimeError>
+where
+    A: BackendApi + 'static,
+    S: Storage + 'static,
+    Q: Querier + 'static,
+{
+    let contract_addr_raw = read_region(&env.memory(), address, 64)?;
+    let contract_addr: Addr = from_slice(&contract_addr_raw)
+        .map_err(|_| RuntimeError::new("Invalid contract address to detect interface"))?;
```
Comment (shiki-tak) [inline packages/vm/src/dynamic_link.rs:251]: It's a matter of word, Are "detect" and "validate" used interchangeably?.
(All other places are validate, but only here is detect.)

Comment (loloicci) [reply packages/vm/src/dynamic_link.rs:251]: nice found. thank you.

File: packages/vm/src/dynamic_link.rs [hunk 3]
```diff
@@ -234,6 +236,56 @@ where
...
+            };
+            err_msg.push_str(&format!("{}: {}", expected_fn.name(), expected_fn.ty()));
+            is_err = true;
+        }
+    }
+
+    if is_err {
+        // not expected
+        Ok(write_to_contract::<A, S, Q>(env, err_msg.as_bytes())?)
+    } else {
```
Comment (shiki-tak) [inline packages/vm/src/dynamic_link.rs:283]: nit: Is there a no problem without else state?

**Review by shiki-tak — APPROVED** [review]
LGTM

**Comment by zemyblue** [comment]
> > How can I check the code coverage?
> 
> now we do not have features to check code coverage...

Could you add codecov checking as CI later?

**Review by zemyblue — APPROVED** [review]

**Review by da1suk8 — APPROVED** [review]

File: packages/vm/src/dynamic_link.rs (hunk 3 above)
Comment (loloicci) [reply packages/vm/src/dynamic_link.rs:283]: L272?

Comment (loloicci) [reply packages/vm/src/dynamic_link.rs:283]: L272?

Comment (loloicci) [reply packages/vm/src/dynamic_link.rs:283]: L272? if it is, there are no problems. It is for adding ',' into the error message's tail before adding a new sentence _if it is not the first sentence_.

Comment (shiki-tak) [reply packages/vm/src/dynamic_link.rs:283]: I thought else statement is unnecessary, but I understood.

Comment (loloicci) [reply packages/vm/src/dynamic_link.rs:283]: The then clause needs to return the same typed value as the else clause. So this part needs to be one of as this PR or

```
if is_err {
    return Ok(write_to_contract...)
};
Ok(0)
```

And, I chose the style as this PR.

**Pull request merged**